### PR TITLE
Release 3.2.1

### PR DESCRIPTION
**Pull request by cronicc**

### CODE REVIEW

File: entrypoint.sh [hunk 1]
```diff
@@ -1,236 +1,227 @@
...
-if ! grep -q 'rpcuser' $HOME/.zen/testnet3/zen.conf ; then
-    echo 'rpcuser=user' >> $HOME/.zen/testnet3/zen.conf
-fi
-if ! grep -q 'rpcpassword' $HOME/.zen/zen.conf ; then
-    echo "rpcpassword=`head -c 32 /dev/urandom | base64`" >> $HOME/.zen/zen.conf
-fi
-if ! grep -q 'rpcpassword' $HOME/.zen/testnet3/zen.conf ; then
-    echo "rpcpassword=`head -c 32 /dev/urandom | base64`" >> $HOME/.zen/testnet3/zen.conf
-fi
+  fi
```
Comment (otoumas) [inline entrypoint.sh:55]: These two if statements could be merged into a single one

File: entrypoint.sh [hunk 2]
```diff
@@ -1,236 +1,227 @@
...
+  ln -fs "${mountpoints[i]}" "${targets[i]}"
+done
+
+# ensure we have minimal mainnet and testnet zen.conf files
+mkdir -p "${HOME}/.zen/testnet3"
+for file in "${HOME}/.zen/zen.conf" "${HOME}/.zen/testnet3/zen.conf"; do
+  touch "${file}"
+  grep -q 'rpcuser' "${file}" || echo 'rpcuser=user' >> "${file}"
+  grep -q 'rpcpassword' "${file}" || echo "rpcpassword=$(head -c 32 /dev/urandom | base64)" >> "${file}"
+done
```
Comment (otoumas) [inline entrypoint.sh:65]: Nice refactor!

**Review by otoumas — APPROVED** [review]
LGTM, added some non blocking comments

**Review by lander86 — APPROVED** [review]
LGTM

File: entrypoint.sh [hunk 3]
```diff
@@ -1,236 +1,227 @@
...
 
 # Prepend some default command line options to OPTS, user provided values will be appended and take precedence.
-export OPTS="-listenonion=0 $OPTS"
+export OPTS="-listenonion=0 ${OPTS:-}"
 
 # Logging to stdout or debug.log
-if [[ -v LOG ]] && [ "$LOG" == "STDOUT" ]; then
-    export OPTS="-printtoconsole $OPTS"
+if [ "${LOG:-}" = "STDOUT" ]; then
+  export OPTS="-printtoconsole $OPTS"
```
Comment (ysibirski) [inline entrypoint.sh:72]: Can probably have curly braces as well as on line #68

File: entrypoint.sh [hunk 4]
```diff
@@ -1,236 +1,227 @@
...
-    fi
-    exec "$@"
-fi
+# convert $OPTS into array
+mapfile -t OPTS < <(sed 's/  */ /g; s/ $//g' <<< "${OPTS}" | tr ' ' '\n')
 
+gosu_cmd=""
+[ "${CURRENT_UID}" -ne 0 ] && gosu_cmd="/usr/local/bin/gosu user"
+if [ "$1" = "zend" ]; then
+  $gosu_cmd zen-fetch-params
```
Comment (ysibirski) [inline entrypoint.sh:222]: Should this be "${gosu_cmd}" as well?

File: entrypoint.sh [hunk 5]
```diff
@@ -1,236 +1,227 @@
...
+  for arg in "${OPTS[@]}"; do
+    set -- "$@" "${arg}"
+  done
+fi
+exec $gosu_cmd "$@"
```
Comment (ysibirski) [inline entrypoint.sh:227]: Same as line #222?

**Review by ysibirski — COMMENTED** [review]
Left you couple of comments @cronicc. Otherwise refactoring looks great like a perfect example of bash best practices. I like it a lot.

One comment I have left may be breaking the desired outcome potentially, but I may be wrong

File: entrypoint.sh (hunk 1 above)
Comment (cronicc) [reply entrypoint.sh]: Done

File: entrypoint.sh (hunk 2 above)
Comment (cronicc) [reply entrypoint.sh:64]: Thanks

File: entrypoint.sh (hunk 3 above)
Comment (cronicc) [reply entrypoint.sh]: Done

File: entrypoint.sh [hunk 6]
```diff
@@ -1,236 +1,227 @@
...
-        else
-            EXTERNAL+=( "$entry" )
-        fi
-    done
+external=()
+if [ -n "${EXTERNAL_IP:-}" ]; then
+  mapfile -t external <<< "$(tr ',' '\n' <<< "${EXTERNAL_IP}")"
+  for i in "${!external[@]}"; do
+    if [ "${external[i]}" = "DETECT" ]; then
+      unset 'external[i]'
```
[the remark this answers is not in the document]
Comment (cronicc) [reply entrypoint.sh:155]: Unset works on var names, not their values, so this is fine.

File: entrypoint.sh (hunk 4 above)
Comment (cronicc) [reply entrypoint.sh:221]: In this case we need `/usr/local/bin/gosu` and  `user` to be interpreted as separate commands. That only works if we word split without double quotes.

**Pull request merged**